apps/s_client.c

```diff
@@ -2257,6 +2259,62 @@ int s_client_main(int argc, char **argv)
             }
         }
         break;
+    case PROTO_MYSQL:
+        {
+            /*
+             * https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_packets.html
+             * https://dev.mysql.com/doc/internals/en/ssl.html
+             * https://dev.mysql.com/doc/internals/en/connection-phase-packets.html#packet-Protocol::HandshakeResponse
+             */
+            static const unsigned char mysql_ssl_request[] = {
+                /* payload_length,   sequence_id */
+                   0x20, 0x00, 0x00, 0x01,
+                /* payload */
+                /* capability flags, CLIENT_SSL always set */
+                   0x85, 0xae, 0x7f, 0x00,
+                /* max-packet size */
+                   0x00, 0x00, 0x00, 0x01,
+                /* character set */
+                   0x21,
+                /* string[23] reserved (all [0]) */
+                   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
+            };
+            int bytes;
+            int ssl_capability = 0x800;
+            char *server_version;
+            int server_version_len;
+
+            /*
+             * https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase.html
+             * Receiving Initial Handshake packet from mysql server
+             */
+            bytes = BIO_read(sbio, sbuf, BUFSIZZ);
+            if (bytes < 1){
+                BIO_printf(bio_err, "BIO_read failed\n");
+                goto shut;
+            } else if (bytes < 4 ||
+                       bytes != (sbuf[0] + sbuf[1]  * 256 + sbuf[2]  * 256 * 256 + 4 ) ) {
+                BIO_printf(bio_err, "MySQL packet length is wrong.\n");
+                goto shut;
+            } else if ( sbuf[4] != 0xA ) {
+                BIO_printf(bio_err, "Only MySQL protocol version 10 is supported.\n");
+                goto shut;
+            } else if( sscanf(&sbuf[5], "%ms", &server_version) != 1 || (server_version_len = strlen(server_version)) < 1 ) {
+                BIO_printf(bio_err, "Cannot confirm server version. MySQL server handshake packet is broken.\n");
+                goto shut;
+            } else if( ! (( sbuf[19 + server_version_len] * 256 + sbuf[20 + server_version_len] ) & ssl_capability ) ) {
```

Comment:
> Is Salt field actually fixed-length? I mean there is trailing zero, which makes you wonder if it's null-terminated [binary] string... I mean if it was fixed-length what do youthey need zero for?

I had thought the same thing, but it seems fixed length per documentation and captured packet. Mysql documentation has moved to new one, and cannot find the explanation in the current documentation, but I confirmed it is fixed length per "Protocol::HandshakeV10" definition:

http://web.archive.org/web/20160814002743/http://dev.mysql.com/doc/internals/en/connection-phase-packets.html#packet-Protocol::Handshake